File: src/main/kotlin/com/zacksimpson/reminders/data/FirestoreValue.kt
package com.zacksimpson.reminders.data

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

/**
 * Converts between plain kotlinx.serialization JSON and Firestore REST API's typed
 * "Value" wire format (each field wrapped as `{"stringValue": "..."}`, etc — see
 * https://firebase.google.com/docs/firestore/reference/rest/v1/Value).
 *
 * Only handles the value types reminders-web actually uses (numbers/strings/bools/
 * arrays/maps) — no timestampValue, referenceValue, geoPointValue, bytesValue.
 */
object FirestoreValue {
    /** A plain JSON field map (e.g. from `Json.encodeToJsonElement(Task.serializer(), t)`)
     *  into Firestore's per-field typed wrapper, ready to send as a document's `fields`. */
    fun encodeFields(fields: JsonObject): JsonObject =
        JsonObject(fields.mapValues { (_, v) -> encode(v) })

    /** The `fields` map from a Firestore document response back into plain JSON,
     *  decodable by `Json.decodeFromJsonElement(Task.serializer(), ...)`. */
    fun decodeFields(fields: JsonObject): JsonObject =
        JsonObject(fields.mapValues { (_, v) -> decode(v.jsonObject) })

    private fun encode(element: JsonElement): JsonObject = when (element) {
        is JsonNull -> buildJsonObject { put("nullValue", JsonNull) }
        is JsonArray -> buildJsonObject {
            put(
                "arrayValue",
                buildJsonObject { put("values", JsonArray(element.map { encode(it) })) },
            )
        }
        is JsonObject -> buildJsonObject {
            put("mapValue", buildJsonObject { put("fields", encodeFields(element)) })
        }
        is JsonPrimitive -> when {
            !element.isString && element.booleanOrNull != null ->
                buildJsonObject { put("booleanValue", element.boolean) }
            // Firestore's wire format encodes integerValue as a numeric *string*, unlike
            // every other numeric field type.
            !element.isString && element.longOrNull != null ->
                buildJsonObject { put("integerValue", element.content) }
            !element.isString ->
                buildJsonObject { put("doubleValue", element.double) }
            else -> buildJsonObject { put("stringValue", element.content) }
        }
    }

    private fun decode(value: JsonObject): JsonElement = when {
        "stringValue" in value -> JsonPrimitive(value.getValue("stringValue").jsonPrimitive.content)
        "booleanValue" in value -> JsonPrimitive(value.getValue("booleanValue").jsonPrimitive.boolean)
        "integerValue" in value -> JsonPrimitive(value.getValue("integerValue").jsonPrimitive.content.toLong())
        "doubleValue" in value -> JsonPrimitive(value.getValue("doubleValue").jsonPrimitive.double)
        "arrayValue" in value -> {
            val values = value["arrayValue"]?.jsonObject?.get("values")?.let { it as? JsonArray } ?: JsonArray(emptyList())
            JsonArray(values.map { decode(it.jsonObject) })
        }
        "mapValue" in value -> {
            val innerFields = value["mapValue"]?.jsonObject?.get("fields")?.jsonObject ?: JsonObject(emptyMap())
            decodeFields(innerFields)
        }
        else -> JsonNull // nullValue, or an unrecognized/absent Value variant.
    }
}

private val JsonPrimitive.boolean: Boolean get() = content.toBoolean()
private val JsonPrimitive.double: Double get() = content.toDouble()

File: utils/sync/firestoreValue.ts
/**
 * Converts between plain JSON and Firestore REST API's typed "Value" wire format (each
 * field wrapped as `{"stringValue": "..."}`, etc, see
 * https://firebase.google.com/docs/firestore/reference/rest/v1/Value).
 *
 * Only handles the value types reminders-web actually uses (numbers/strings/bools/
 * arrays/maps/null), no timestampValue, referenceValue, geoPointValue, bytesValue.
 */

export type JsonValue =
  | boolean
  | JsonValue[]
  | null
  | number
  | string
  | { [key: string]: JsonValue };

type FirestoreValue = Record<string, unknown>;

/** A plain JSON field map into Firestore's per-field typed wrapper, ready to send as a
 *  document's `fields`. */
export function encodeFields(
  fields: Record<string, JsonValue | undefined>
): Record<string, FirestoreValue> {
  const out: Record<string, FirestoreValue> = {};
  for (const [key, value] of Object.entries(fields)) {
    out[key] = encodeValue(value);
  }
  return out;
}

function encodeValue(value: JsonValue | undefined): FirestoreValue {
  // A field explicitly set to undefined (e.g. an optional Task field like `date`)
  // encodes the same as an absent/null one, Firestore has no "undefined" value type.
  if (value === null || value === undefined) {
    return { nullValue: null };
  }
  if (Array.isArray(value)) {
    return { arrayValue: { values: value.map(encodeValue) } };
  }
  if (typeof value === "object") {
    return { mapValue: { fields: encodeFields(value) } };
  }
  if (typeof value === "boolean") {
    return { booleanValue: value };
  }
  if (typeof value === "number") {
    // Firestore's wire format encodes integerValue as a numeric *string*, unlike every
    // other numeric field type.
    return Number.isInteger(value)
      ? { integerValue: String(value) }
      : { doubleValue: value };
  }
  return { stringValue: value };
}

/** The `fields` map from a Firestore document response back into plain JSON. */
export function decodeFields(
  fields: Record<string, unknown>
): Record<string, JsonValue> {
  const out: Record<string, JsonValue> = {};
  for (const [key, value] of Object.entries(fields)) {
    out[key] = decodeValue(value as FirestoreValue);
  }
  return out;
}

function decodeValue(value: FirestoreValue): JsonValue {
  if (!value) {
    return null;
  }
  if ("stringValue" in value) {
    return value.stringValue as string;
  }
  if ("booleanValue" in value) {
    return value.booleanValue as boolean;
  }
  if ("integerValue" in value) {
    return Number(value.integerValue);
  }
  if ("doubleValue" in value) {
    return value.doubleValue as number;
  }
  if ("arrayValue" in value) {
    const values =
      ((value.arrayValue as { values?: FirestoreValue[] } | undefined)
        ?.values ?? []) as FirestoreValue[];
    return values.map(decodeValue);
  }
  if ("mapValue" in value) {
    const innerFields =
      (value.mapValue as { fields?: Record<string, unknown> } | undefined)
        ?.fields ?? {};
    return decodeFields(innerFields);
  }
  return null; // nullValue, or an unrecognized/absent Value variant.
}
